# Jupyter Notebook: Data cleaning

## Step 1: Install Required Libraries

First, ensure you have the necessary libraries installed. You can install them using the following command:

`pip install pandas numpy seaborn matplotlib`

## Step 2: Import Required Libraries


In [5]:
from multiprocessing.reduction import duplicate

import pandas as pd
import numpy as np
import os

## Step 3: Load the titanic dataset
Update the variable file_path to the path where you have your data.

In [6]:
df = pd.read_csv(os.path.join('titanic.csv'))

# Visualizing the first few rows of the dataset
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Step 4: Understanding the Data
Check dataset information, including data types and non-null counts
This helps us identify missing values and understand the data format

Check for missing values in each column to plan how to handle them

In [7]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          688
Embarked         2
dtype: int64

## Step 5: Removing missing values
For numerical columns (e.g., Age) we can fill missing values with the median.

In [8]:
# Fill missing 'Age' values with the median
df['Age'] = df["Age"].fillna(df['Age'].median())
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


# Step 6: Remove irrelevant data
This dataset is used to predict if a passenger would survive the Titanic based on the information that we have about that passenger. However, some of the columns of the dataset contain information which is irrelevant to predict a passenger's chance of surviving. For example the passengner ID.

In [9]:
df.drop(['PassengerId'], axis=1, inplace=True)

df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


# Step 7: Convert categorical values to numerical
Some of the columns of the dataset contain categorical information like Sex and Embarked. However, machine learning algorithms require numerical input. We are going to transform the categorical values from the Sex column (female and male) to numerical by assigning a 0 to male and 1 to female. 

In [10]:
# convert sex to numerical
df['Sex'] = df['Sex'].replace(['male', 'female'], [0, 1])

df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,A/5 21171,7.2500,NaN,S
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,PC 17599,71.2833,C85,C
2,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,113803,53.1000,C123,S
4,0,3,"Allen, Mr. William Henry",0,35.0,0,0,373450,8.0500,NaN,S


### Exercise 1

Determine what percentage of the cabin data is missing. Explain and justify what you think it should be done to handle those missing values. Then, write the corresponding code. 

<div style="text-align: right;">
  (1 mark)
</div>

<span style="color:red">Your answer goes here</span>

In [11]:
# Your answer goes here
# Amount of missing data
df.isnull().sum()["Cabin"] / len(df)

np.float64(0.7713004484304933)

Therefore we are missing 77.13%. Given that we are missing so much of the data, i think it would be reasonable to ignore it for model training, as there is so much missing data that it is not possible to train a model off of such lackluster data.

In [12]:
df = df.drop(columns=["Cabin"])
df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,A/5 21171,7.2500,S
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,PC 17599,71.2833,C
2,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,STON/O2. 3101282,7.9250,S
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,113803,53.1000,S
4,0,3,"Allen, Mr. William Henry",0,35.0,0,0,373450,8.0500,S


### Exercise 2

Idenfity and justify which columns are irrelevant to determine if a passenger will survive the Titanic. Then, write the corresponding code. 

<div style="text-align: right;">
  (1 mark)
</div>

<span style="color:red">Your answer goes here</span>

1. Ticket is not relevant, Ticket id cannot be used to predict or use anything with it
2. Cost does not matter, We already have cabin class -> therefore we do not need cost

In [13]:
df.drop(['Fare'], axis=1, inplace=True)
df.drop(['Ticket'], axis=1, inplace=True)
df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Embarked
0,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,S
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,C
2,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,S
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,S
4,0,3,"Allen, Mr. William Henry",0,35.0,0,0,S


### Exercise 3

Transform the Embarked column from categorical to numerical.

<div style="text-align: right;">
  (1 mark)
</div>

In [14]:
df['Embarked'] = df['Embarked'].replace(['S', 'C','Q'], [0, 1, 2])
df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Embarked
0,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,0
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,1
2,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,0
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,0
4,0,3,"Allen, Mr. William Henry",0,35.0,0,0,0


### Exercise 4

Identify what type of data is the Embarked column. Explain why you can't fill in the missing values with the mean. Propose and justify an alternative and provide the corresponding code. 

<div style="text-align: right;">
  (2 marks)
</div>

The data in the Embarked column is Categorical data, because it is so it cannot be replaced with a mean of different categorical data due to you being unable to find the average of categorical data.

However, you can find the most common data and replace the missing with the mode.

In [15]:
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])
df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Embarked
0,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,0
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,1
2,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,0
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,0
4,0,3,"Allen, Mr. William Henry",0,35.0,0,0,0


### Exercise 5

Examine the Age column to identify any values that are impossible or unrealistic (e.g., negative ages or extreme values). Handle them appropriately.

<div style="text-align: right;">
  (2 marks)
</div>

In [31]:
mean_age = df['Age'].mean()

df.loc[(df['Age'] < 0) | (df['Age'] > 100), 'Age'] = mean_age


### Exercise 6

Calculate the percentage of passengers who survived the Titanic disaster.
(Hint: search for the `value_counts` function in the Pandas documentation)

<div style="text-align: right;">
  (1 mark)
</div>

In [17]:
df['Survived'].mean() * 100

np.float64(38.34080717488789)

### Exercise 7

Check whether the dataset contains any duplicate rows. If duplicates are found, handle them appropriately.

<div style="text-align: right;">
  (1 mark)
</div>

In [29]:
df.duplicated().sum()
df = df.drop_duplicates()
df

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Embarked
0,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,0
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,1
2,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,0
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,0
4,0,3,"Allen, Mr. William Henry",0,35.0,0,0,0
...,...,...,...,...,...,...,...,...
886,0,2,"Montvila, Rev. Juozas",0,27.0,0,0,0
887,1,1,"Graham, Miss. Margaret Edith",1,19.0,0,0,0
888,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",1,28.0,1,2,0
889,1,1,"Behr, Mr. Karl Howell",0,26.0,0,0,1


### Exercise 8

Previously, we filled missing age values using the average age. Propose two alternative methods for handling these missing values. For each method, explain how it might influence the results of future statistical analyses.

<div style="text-align: right;">
  (2 marks)
</div>

Instead of inserting the average age, we have two more options to resolve such a dillema.
1. Insert the median, doing would result in a less influence on the average from outliers. Also doing so would remove the graph from being biased towards the older age demographics.
2. A better way is to use the median within groups, like via gender or class. Preferably class as it would result in more relable data. Doing so results in better relationship between variables, like class.

### Exercise 9

Examine the Embarked column to determine whether the data is consistently formatted (for example, consistent use of uppercase and lowercase letters). If inconsistencies exist, clean the column to ensure uniform formatting.

<div style="text-align: right;">
  (2 marks)
</div>

In [26]:
df["Embarked"].unique()
df['Embarked'] = df['Embarked'].replace(['s', 'c','q'], [0, 1, 2])
df
df["Embarked"].unique()

array([0, 1, 2], dtype=object)

### Exercise 10

Save the cleaned dataset in a new csv file. Be careful not to overwrite the original file.

<div style="text-align: right;">
  (1 mark)
</div>

In [28]:
df.to_csv("cleaned_dataset.csv", index=False)

## Extension 1

Create a new column called FamilySize by combining SibSp (siblings/spouses) and Parch (parents/children).


### Extension 2

Extract passenger titles from the Name column (e.g., Mr., Mrs., Miss.). Investigate whether there are any inconsistencies between a passenger’s title and their recorded gender or age, and describe any issues you find.